File: Windows/FireflyDevice/FDHelloTask.cpp
//
//  FDHelloTask.cpp
//  FireflyDevice
//
//  Created by Denis Bohm on 10/6/13.
//  Copyright (c) 2013-2014 Firefly Design LLC / Denis Bohm. All rights reserved.
//

#include "FDError.h"
#include "FDFireflyIceChannel.h"
#include "FDFireflyIceCoder.h"
#include "FDFireflyDeviceLogger.h"
#include "FDHelloTask.h"
#include "FDTime.h"

namespace FireflyDesign {

	FDHelloTask::FDHelloTask(std::shared_ptr<FDFireflyIce> fireflyIce, std::shared_ptr<FDFireflyIceChannel> channel, std::shared_ptr<FDHelloTaskDelegate> delegate)
		: FDFireflyIceTaskSteps(fireflyIce, channel)
	{
		this->delegate = delegate;
		priority = 100;
		_maxOffset = 120;
		_time = 0;
	}

	void FDHelloTask::fireflyIceVersion(std::shared_ptr<FDFireflyIce> fireflyIce, std::shared_ptr<FDFireflyIceChannel> channel, FDFireflyIceVersion version) {
		_version = std::make_shared<FDFireflyIceVersion>(version);
	}

	void FDHelloTask::fireflyIceHardwareId(std::shared_ptr<FDFireflyIce> fireflyIce, std::shared_ptr<FDFireflyIceChannel> channel, FDFireflyIceHardwareId hardwareId) {
		_hardwareId = std::make_shared<FDFireflyIceHardwareId>(hardwareId);
	}

	void FDHelloTask::fireflyIceBootVersion(std::shared_ptr<FDFireflyIce> fireflyIce, std::shared_ptr<FDFireflyIceChannel> channel, FDFireflyIceVersion bootVersion) {
		_bootVersion = std::make_shared<FDFireflyIceVersion>(bootVersion);
	}

	void FDHelloTask::fireflyIceTime(std::shared_ptr<FDFireflyIce> fireflyIce, std::shared_ptr<FDFireflyIceChannel> channel, time_type time) {
		_time = time;
	}

	void FDHelloTask::fireflyIcePower(std::shared_ptr<FDFireflyIce> fireflyIce, std::shared_ptr<FDFireflyIceChannel> channel, FDFireflyIcePower power) {
		_power = std::make_shared<FDFireflyIcePower>(power);
	}

	void FDHelloTask::fireflyIceReset(std::shared_ptr<FDFireflyIce> fireflyIce, std::shared_ptr<FDFireflyIceChannel> channel, FDFireflyIceReset reset) {
		_reset = std::make_shared<FDFireflyIceReset>(reset);
	}

	void FDHelloTask::executorTaskStarted(FDExecutor *executor)
	{
		FDFireflyIceTaskSteps::executorTaskStarted(executor);

		fireflyIce->coder->sendGetProperties(channel,
			FD_CONTROL_PROPERTY_VERSION |
			FD_CONTROL_PROPERTY_HARDWARE_ID |
			FD_CONTROL_PROPERTY_RTC |
			FD_CONTROL_PROPERTY_POWER |
			FD_CONTROL_PROPERTY_RESET |
			FD_CONTROL_PROPERTY_BOOT_VERSION
		);
		next(std::bind(&FDHelloTask::checkVersion, this));
	}

	void FDHelloTask::checkVersion()
	{
		if (!_version || !_hardwareId) {
			std::string description = "Incomplete information received on initial communication with the device";
			FDFireflyDeviceLogInfo(description);
			channel->close();
			std::map<std::string, std::string> userInfo;
			userInfo[FDLocalizedDescriptionKey] = description;
			userInfo[FDLocalizedRecoveryOptionsErrorKey] = "Make sure the device stays in close range";
			std::shared_ptr<FDError> error = FDError::error(FDHelloTaskErrorDomain, FDHelloTaskErrorCodeIncomplete, userInfo);
			fireflyIce->executor->fail(shared_from_this(), error);
			return;
		}

		fireflyIce->version = _version;
		fireflyIce->bootVersion = _bootVersion;
		fireflyIce->hardwareId = _hardwareId;

		if (fireflyIce->version->capabilities & FD_CONTROL_CAPABILITY_BOOT_VERSION) {
			fireflyIce->coder->sendGetProperties(channel, FD_CONTROL_PROPERTY_BOOT_VERSION);
			next(std::bind(&FDHelloTask::checkTime, this));
		} else {
			checkTime();
		}
	}

	void FDHelloTask::setTime()
	{
		FDTime::time_type time = delegate != NULL ? delegate->helloTaskDate() : FDTime::time();
		if (time != 0) {
			FDFireflyDeviceLogInfo("setting the time");
			fireflyIce->coder->sendSetPropertyTime(channel, time);
		}
	}

	void FDHelloTask::checkTime()
	{
		std::string hardwareIdDescription = fireflyIce->hardwareId->description();
		std::string versionDescription = fireflyIce->version->description();
		std::string resetDescription = _reset->description();
		FDFireflyDeviceLogInfo("hello (hardware %s) (firmware %s)", hardwareIdDescription.c_str(), versionDescription.c_str());

		if (_time == 0) {
			FDFireflyDeviceLogInfo("time not set for hw %s fw %s (last reset %s)", hardwareIdDescription.c_str(), versionDescription.c_str(), resetDescription.c_str());
			setTime();
		} else {
			FDTime::time_type time = delegate != NULL ? delegate->helloTaskDate() : FDTime::time();
			if (time != 0) {
				duration_type offset = time - _time;
				if (fabs(offset) > _maxOffset) {
					FDFireflyDeviceLogInfo("time is off by %0.3f seconds for hw %s fw %s (last reset %s)", offset, hardwareIdDescription.c_str(), versionDescription.c_str(), resetDescription.c_str());
					setTime();
				} else {
					//            FDFireflyDeviceLogDebug(@"time is off by %0.3f seconds for hw %@ fw %@", offset, self.fireflyIce.hardwareId, self.fireflyIce.version);
				}
			}
		}
		done();
	}

	void FDHelloTask::executorTaskCompleted(FDExecutor *executor)
	{
		FDFireflyIceTaskSteps::executorTaskCompleted(executor);
		if (delegate != nullptr) {
			delegate->helloTaskSuccess(this);
		}
	}

	void FDHelloTask::executorTaskFailed(FDExecutor *executor, std::shared_ptr<FDError> error)
	{
		FDFireflyIceTaskSteps::executorTaskFailed(executor, error);
		if (delegate != nullptr) {
			delegate->helloTaskError(this, error);
		}
	}

}

File: Android/Libraries/fireflydevice/src/main/java/com/fireflydesign/fireflydevice/FDHelloTask.java
//
//  FDHelloTask.cpp
//  FireflyDevice
//
//  Created by Denis Bohm on 10/6/13.
//  Copyright (c) 2013-2014 Firefly Design LLC / Denis Bohm. All rights reserved.
//

package com.fireflydesign.fireflydevice;

import java.util.HashMap;
import java.util.Map;

public class FDHelloTask extends FDFireflyIceTaskSteps {

    public final static String FDHelloTaskErrorDomain = "com.fireflydesign.device.FDHelloTask";

    public enum ErrorCode {
        Incomplete
    }

    public interface Delegate {
        double helloTaskDate();
        void helloTaskSuccess(FDHelloTask helloTask);
        void helloTaskError(FDHelloTask helloTask, FDError error);
    }

    public Delegate delegate;
    public FDFireflyIceVersion version;
    public FDFireflyIceVersion bootVersion;
    public FDFireflyIceHardwareId hardwareId;
    public double time;
    public FDFireflyIcePower power;
    public FDFireflyIceReset reset;

    double maxOffset;

    public FDHelloTask(FDFireflyIce fireflyIce, FDFireflyIceChannel channel, Delegate delegate) {
        super(fireflyIce, channel);
		this.delegate = delegate;
		priority = 100;
		maxOffset = 120;
		time = 0;
	}

	public void fireflyIceVersion(FDFireflyIce fireflyIce, FDFireflyIceChannel channel, FDFireflyIceVersion version) {
		this.version = version;
	}

    public void fireflyIceHardwareId(FDFireflyIce fireflyIce, FDFireflyIceChannel channel, FDFireflyIceHardwareId hardwareId) {
		this.hardwareId = hardwareId;
	}

    public void fireflyIceBootVersion(FDFireflyIce fireflyIce, FDFireflyIceChannel channel, FDFireflyIceVersion bootVersion) {
		this.bootVersion = bootVersion;
	}

    public void fireflyIceTime(FDFireflyIce fireflyIce, FDFireflyIceChannel channel, double time) {
		this.time = time;
	}

    public void fireflyIcePower(FDFireflyIce fireflyIce, FDFireflyIceChannel channel, FDFireflyIcePower power) {
		this.power = power;
	}

    public void fireflyIceReset(FDFireflyIce fireflyIce, FDFireflyIceChannel channel, FDFireflyIceReset reset) {
		this.reset = reset;
	}

    public void executorTaskStarted(FDExecutor executor) {
		super.executorTaskStarted(executor);

		fireflyIce.coder.sendGetProperties(
            channel,
			FDFireflyIceCoder.FD_CONTROL_PROPERTY_VERSION |
            FDFireflyIceCoder.FD_CONTROL_PROPERTY_HARDWARE_ID |
            FDFireflyIceCoder.FD_CONTROL_PROPERTY_RTC |
            FDFireflyIceCoder.FD_CONTROL_PROPERTY_POWER |
            FDFireflyIceCoder.FD_CONTROL_PROPERTY_RESET |
            FDFireflyIceCoder.FD_CONTROL_PROPERTY_BOOT_VERSION
		);
		next("checkVersion");
	}

    public void executorTaskCompleted(FDExecutor executor) {
        super.executorTaskCompleted(executor);
        if (delegate != null) {
            delegate.helloTaskSuccess(this);
        }
    }

    public void executorTaskFailed(FDExecutor executor, FDError error) {
        super.executorTaskFailed(executor, error);
        if (delegate != null) {
            delegate.helloTaskError(this, error);
        }
    }

    void checkVersion() {
		if ((version == null) || (hardwareId == null)) {
			String description = "Incomplete information received on initial communication with the device";
			FDFireflyDeviceLogger.info(log, "FD010501", description);
			channel.close();
			Map<String, String> userInfo = new HashMap<String, String>();
			userInfo.put(FDError.FDLocalizedDescriptionKey, description);
			userInfo.put(FDError.FDLocalizedRecoveryOptionsErrorKey, "Make sure the device stays in close range");
			FDError error = FDError.error(FDHelloTaskErrorDomain, ErrorCode.Incomplete.ordinal(), userInfo);
			fireflyIce.executor.fail(this, error);
			return;
		}

		fireflyIce.version = version;
		fireflyIce.bootVersion = bootVersion;
		fireflyIce.hardwareId = hardwareId;

		if ((fireflyIce.version.capabilities & FDFireflyIceCoder.FD_CONTROL_CAPABILITY_BOOT_VERSION) != 0) {
			fireflyIce.coder.sendGetProperties(channel, FDFireflyIceCoder.FD_CONTROL_PROPERTY_BOOT_VERSION);
			next("checkTime");
		} else {
			checkTime();
		}
	}

	void setTime() {
		double time = (delegate != null) ? delegate.helloTaskDate() : FDTime.time();
		if (time != 0) {
			FDFireflyDeviceLogger.info(log, "FD010502", "setting the time");
			fireflyIce.coder.sendSetPropertyTime(channel, time);
		}
	}

	void checkTime() {
		String hardwareIdDescription = fireflyIce.hardwareId.description();
		String versionDescription = fireflyIce.version.description();
		String resetDescription = reset.description();
		FDFireflyDeviceLogger.info(log, "FD010503", "hello (hardware %s) (firmware %s)", hardwareIdDescription, versionDescription);

		if (time == 0) {
			FDFireflyDeviceLogger.info(log, "FD010504", "time not set for hw %s fw %s (last reset %s)", hardwareIdDescription, versionDescription, resetDescription);
			setTime();
		} else {
			double time = (delegate != null) ? delegate.helloTaskDate() : FDTime.time();
			if (time != 0) {
				double offset = time - this.time;
				if (Math.abs(offset) > maxOffset) {
					FDFireflyDeviceLogger.info(log, "FD010505", "time is off by %.3f seconds for hw %s fw %s (last reset %s)", offset, hardwareIdDescription, versionDescription, resetDescription);
					setTime();
				} else {
					FDFireflyDeviceLogger.info(log, "FD010506", "time is off by %.3f seconds for hw %s fw %s", offset, hardwareIdDescription, versionDescription);
				}
			}
		}
		done();
	}

}
